### PR TITLE
[RF] Add overloads for memory-safe `RooAbsCollection::addOwned` and `RooAbsArg::addOwnedComponents`

### PR DESCRIPTION
**Pull request by guitargeek**
More details in in the commit descriptions.

In particular, the memory management in JSONFactories_HistFactory is simplified in such a way that this PR should unblock https://github.com/root-project/root/pull/9683.

### CODE REVIEW

File: roofit/roofitcore/src/RooAbsArg.cxx [hunk 1]
```diff
@@ -2275,6 +2275,27 @@ Bool_t RooAbsArg::addOwnedComponents(const RooArgSet& comps)
...
+bool RooAbsArg::addOwnedComponents(RooAbsCollection&& comps)
+{
+  if (!_ownedComponents) {
+    _ownedComponents = new RooArgSet("owned components") ;
+  }
+  return _ownedComponents->addOwned(std::move(comps)) ;
+}
+
+
+bool RooAbsArg::addOwnedComponents(const RooArgList& comps) {
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooAbsArg.cxx:2292]: Add doxygen.

Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsArg.cxx:2292]: You can use `\copydoc` if you just want to copy the docs from above.

File: roofit/roofitcore/src/RooAbsCollection.cxx [hunk 2]
```diff
@@ -458,6 +458,16 @@ Bool_t RooAbsCollection::addOwned(RooAbsArg& var, Bool_t silent)
 }
 
 
+bool RooAbsCollection::addOwned(std::unique_ptr<RooAbsArg> var, bool silent) {
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooAbsCollection.cxx:461]: Doxygen.

File: roofit/roofitcore/src/RooAbsCollection.cxx [hunk 3]
```diff
@@ -514,9 +524,17 @@ Bool_t RooAbsCollection::add(const RooAbsArg& var, Bool_t silent)
 /// Add a collection of arguments to this collection by calling addOwned()
 /// for each element in the source collection
 
-Bool_t RooAbsCollection::addOwned(const RooAbsCollection& list, Bool_t silent)
+bool RooAbsCollection::addOwned(const RooAbsCollection& list, bool silent)
 {
-  Bool_t result(false) ;
+  if(list.isOwning()) {
+    throw std::runtime_error("Passing an owning RooAbsCollection by const& to"
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooAbsCollection.cxx:530]: Suggest `std::invalid_argument` when you get an invalid argument.

**Review by hageboeck — COMMENTED** [review]
This is looking quite good already. See some suggestions below.

**Review by lmoneta — APPROVED** [review]
Thank you Jonas for these very nice improvements !
I think the PR is ready to be merged after addressing the comments from Stephan.

File: roofit/roofitcore/src/RooAbsCollection.cxx (hunk 3 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsCollection.cxx]: Thanks, good suggestion! I have updated it.

File: roofit/roofitcore/src/RooAbsCollection.cxx (hunk 2 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsCollection.cxx:476]: Added!

File: roofit/roofitcore/src/RooAbsArg.cxx (hunk 1 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsArg.cxx]: Right, I have added the doxygen now. Sorry for not adding it before, I first wanted to see the CI tests pass before writing the docs

**Comment by guitargeek** [comment]
The PR was updated with Stephan's suggestions, in particular writing the function documentation.

**Pull request merged**